syntax = "proto3";

import "distribution.proto";
import "mixing_state.proto";
import "matrix.proto";

package bayesmix;


/*
 * Prior for the concentration parameter of a Dirichlet process
 */
message DPPrior {
  message GammaPrior {
    GammaDistribution totalmass_prior = 1;
  }

  oneof totalmass {
    DPState fixed_value = 1; // No prior, just a fixed value
    GammaPrior gamma_prior = 2; // Gamma prior on the total mass
  }
}
/*
 * Prior for the Poisson rate and Dirichlet parameters of a MFM (Finite Dirichlet) process.
 * For the moment, we only support fixed values
 */
message MFMPrior {
  oneof totalmass {
    MFMState fixed_value = 1; // No prior, just a fixed value
  }
}

/*
 * Prior for the strength and discount parameters of a Pitman-Yor process.
 * For the moment, we only support fixed values
 */
message PYPrior {
  oneof totalmass {
    PYState fixed_values = 1;
  }
}

/*
 * Definition of the parameters of a Logit-Stick Breaking process.
 */
message LogSBPrior {
  oneof coeff {
    MultiNormalDistribution normal_prior = 1; // Normal prior on the regression coefficients
  }
  double step_size = 2;  // Steps size for the MALA algorithm used for posterior inference (TODO: move?)
  uint32 num_components = 3; // Number of components in the process
}

/*
 * Definition of a generic container for the prior parameters to be used in Python
 */
message PythonMixPrior {
  oneof prior {
    Vector values = 1;
  }
}

/*
 * Definition of the parameters of a truncated Stick-Breaking process
 */
message TruncSBPrior {
  message BetaPriors {
    // General stick-breaking distributions
    repeated BetaDistribution beta_distributions = 1;
  }
  message DPPrior {
    // Truncated Dirichlet process
    double totalmass = 1;
  }
  message MFMPrior {
    // Truncated Dirichlet process
    double totalmass = 1;
  }
  message PYPrior {
    // Truncated Pitman-Yor process
    double strength = 1;
    double discount = 2;
  }

  oneof prior {
    BetaPriors beta_priors = 1; // General stick-breaking distributions
    DPPrior dp_prior = 2; // Truncated Dirichlet process
    PYPrior py_prior = 3; // Truncated Pitman-Yor process
    MFMPrior mfm_prior = 5;
    PythonMixPrior pm_prior = 6;
  }
  uint32 num_components = 4; // Number of components in the process
}
